p2panda-net/src/sync/manager.rs

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use anyhow::Result;
+use iroh_gossip::proto::TopicId;
+use iroh_net::{Endpoint, NodeId};
+use p2panda_sync::traits::SyncProtocol;
+use p2panda_sync::SyncError;
+use tokio::sync::mpsc::Sender;
+use tracing::{debug, warn};
+
+use crate::engine::ToEngineActor;
+use crate::sync::{self, SYNC_CONNECTION_ALPN};
+
+#[derive(Debug)]
+pub struct SyncManager {
+    known_peer_topics: HashMap<NodeId, HashSet<TopicId>>,
+    active_sync_sessions: HashMap<TopicId, HashSet<NodeId>>,
+    completed_sync_sessions: HashMap<TopicId, HashSet<NodeId>>,
+    engine_actor_tx: Sender<ToEngineActor>,
+    endpoint: Endpoint,
+    sync_protocol: Arc<dyn for<'a> SyncProtocol<'a> + 'static>,
+}
+
+impl SyncManager {
+    pub fn new(
+        endpoint: Endpoint,
+        engine_actor_tx: Sender<ToEngineActor>,
+        sync_protocol: Arc<dyn for<'a> SyncProtocol<'a> + 'static>,
+    ) -> Self {
+        Self {
+            known_peer_topics: HashMap::new(),
+            active_sync_sessions: HashMap::new(),
+            completed_sync_sessions: HashMap::new(),
+            engine_actor_tx,
+            endpoint,
+            sync_protocol,
+        }
+    }
+
+    /// Respond to newly discovered peer topics by initiating a new connection if one is not
+    /// currently underway and a successful sync session has not already been completed.
+    pub async fn update_peer_topics(&mut self, peer: NodeId, topics: Vec<TopicId>) -> Result<()> {
+        debug!("updating peer topics in connection manager");
+
+        // Create a list of (previously) unknown topics that we might like to sync over.
+        let mut new_topics = Vec::new();
```

Comment:
This isn't the most elegant solution (compared to what @sandreae had before) but I needed to ensure we didn't have two mutable borrows.